### ISSUE TITLE
Site configuration improvement

### ISSUE THREAD

**Issue by mmzeeman** [zotonic/zotonic#1183]
It would be really nice if the site configuration could be stored in a different location separate from its git repository. 

What would be a good location for this?

 `.zotonic/<zotonic-version-or-nodename>/sitename.config` or even `.zotonic/sitename.config`?

Or something entirely different? 

This would also be a better location to store ssl certificates and keys.

**Comment by mworrell** [comment]
I think that we might want to move some directories (and/or change some defaults) so that we don't put any user data in any directory that is from Git. In such a way we can make the Git directories read-only and easy to change or replace completely.

All user data (and site configs) should then be concentrated on a single place from where it is easy to upgrade/move/backup sites between Zotonic versions and servers.

For the SSL certificates we might want to add either a separate store, or let the system scan all site directories to create a store where we can map a hostname to a certificate (and a self signed one from the system if no cert is found for that hostname).

**Comment by mworrell** [comment]
Preliminary set to the 1.0 milestone - as this is quite important for operations (and the cowboy SSL/SNI implementation)

**Comment by ddeboer** [comment]
Wait a second, there’s already `sitename/config.d/file`, right? You can just gitignore the `config.d` dir.

**Comment by mmzeeman** [comment]
I’d like to be able to upgrade a site by simply replacing site and module directories without having to take care of config files and the ssl directory containing keys and certificates. 

That can be easily done in .zotonic/nodename/zotonic.config

Currently I’m busy with an installable site which will be managed by others. They will not install and upgrade via git, but from a pre-compiled tarball.

**Comment by mworrell** [comment]
Indeed - the goal is to just be able to do a brute force replace of all site directories (including the Zotonic directory itself).

Maybe @CyBeRoni has some ideas as well?

**Comment by CyBeRoni** [comment]
Keeping it separate is a good idea I think, and it seems to me like a configuration file in the same places as the current zotonic.config and erlang.config named sitename.config would work fine.

I also think that in most cases, it should suffice to not even have any site-specific configuration files unless you need to override something. I.e., use sensible defaults and allow configuring certain behaviours globally. For instance, a global setting to 'use sitename as default database name' and then if for some reason a specific site needs something different, override it there.

Perhaps such a thing could be implemented by allowing some sort of variable replacement in config settings? So for example I could set the default site url to '<sitename>.dev' in my development config files, and the default database to just '<sitename>' and the default schema to 'public'.

**Comment by mworrell** [comment]
I am totally for less configuration.

Currently the really used configuration keys in the site config are:

 * hostname  (could be derived from the site name, could also be in the database)
 * hostalias (could be saved in the database)
 * database, schema etc. (could come from zotonic config + site-name =:= schema-name)
 * admin password (could be saved in the database or even fallback to zotonic config)
 * install_modules (etc), this is static and need to be in a git-based config file anyway

The database settings could also be a zotonic-system database, where we can manage the individual sites. I can see quite some good things for such a system database, like storing system related information, settings and status. This can then also be accessed using SQL tools.

Putting more in the database makes us of course more reliant on the database, but also more flexible.
(A fallback to this system database could be mnesia, though it is not very system-operations-friendly...)

**Comment by CyBeRoni** [comment]
Hostname is probably the trickiest part to deal with here, also because it has security implications in a multi-tenant set-up. You don't want tenant A to be able to hijack tenant B's website by just adding tenant B's url to the tenant A website.

What do you think of the idea of config setting variable expansion?

**Comment by mworrell** [comment]
The primary hostnames could also be configured in the 'zotonic_status' part, overruling any other settings. Right now any tenant can also hijack "tertiary" hostnames by using the fallback handling (that is after the hostname and hostalias resolution).

The config setting variable expansion sounds good to me :-)  Isn't it also some kind of postfix for site names? Much like the search domain for DNS lookups?  I can't really see the use of any infix patterns.

**Comment by CyBeRoni** [comment]
Well, no; Suppose I keep stuff really simple and name my sites after their domain name, I could just say `hostname: "www.${sitename}.com", hostaliases: "${sitename}.com"`.

Come to think of it, does Zotonic do wildcard matching?

**Comment by mworrell** [comment]
No wildcard matching at the moment, though we could add some as the hostname lookups are now done via an ets table, so quite fast.

We could rewrite "foobar.example.com" to "*.example.com" and check the ets table for such an entry.

**Comment by CyBeRoni** [comment]
By the way this 'nuke the site dir' thing also affects uploaded files if you don't use s3 for those.

**Comment by mworrell** [comment]
Indeed, so it is a bit of discussion what we place where. Questions are:

 * Should the .git checkouts be left as-is without any configuration/files etc (.beam and transient files should be ok)
 * If left as-is, where do we store:
   - site configs
  - uploaded files
  - SSL certificates
  - log files
  - mnesia databases (if any)

I am open to suggestions, and maybe pointers to *best practice* solutions in the Unix/BSD/Linux worlds (which are our primary hosting envs).

**Comment by CyBeRoni** [comment]
The actual location is going to be different on each platform and likely even between different variants of the same platform (i.e., Debian vs. CentOS) or whether or not you're using containers or not and if so what flavour (Docker, SmartOS, Rkt, etc.) That is fine, so long as there's a reasonable way to separate the data from the code and tell zotonic what sits where.

So basically each of those things should have a path derived from a configuration setting (with a sensible default). That configuration should be kept in default locations too modifiable with environment variables (already working since #1082). The end result should be that I should be able to wipe the entire system clean, put in a zotonic checkout, site checkout and the relevant configs which point at the persistent data and then everything is back to where it was. (Ideally btw I could also do template pre-compilation but that is another issue.) This sort of thing enables properly containerised applications where you deploy an image consisting of code and config, and then have data on a attachable persistent volume of some sort.

**Comment by mmzeeman** [comment]
Unix distro's usually have separate locations for configuration and data stuff. Configuration stuff usually goes to `/etc` and data to `/var`. 

We also should make that difference I guess. Site configs and ssl stuff is config data, mnesia dbs, logs and uploaded & processed files is data.

**Comment by mworrell** [comment]
Shouldn't we just keep the Zotonic stuff local to the user running Zotonic?
(As long we don't use packages, that should be ok).

**Comment by CyBeRoni** [comment]
What do you mean by 'local to the user running Zotonic'? Put it in the zotonic home dir? Because many people won't have one of those.

I like putting my data in /srv, which is what that dir exists for.

**Comment by mworrell** [comment]
Quite some people run Zotonic from an user account.
So basically you can have two scenarios:

 1. Run from user account
 2. Run from system installation (installed via .deb or otherwise)

If running from user account I would expect most (if not all?) files to be local to the user account. This must be supported, as it is the way to run Zotonic without any changes that need *sudo* permissions.

If running from system installation then I would expect data to be in `/srv/zotonic`, `/var/log/zotonic`, `/etc/zotonic/` and maybe some separate directory for data (`/var/zotonic`) and other system state files.

Just like @mmzeeman said above.

**Comment by CyBeRoni** [comment]
> This must be supported

Absolutely, but it shouldn't be the only way supported. Running from a regular login account isn't exactly best practice for production services. For development it's fine.
